src/lib/logstore/log_store.cpp

```diff
@@ -83,7 +84,10 @@ void HomeLogStore::write_async(logstore_req* req, const log_req_comp_cb_t& cb) {
     HS_LOG_ASSERT((cb || m_comp_cb), "Expected either cb is not null or default cb registered");
     req->cb = (cb ? cb : m_comp_cb);
     req->start_time = Clock::now();
-    if (req->seq_num == 0) { m_safe_truncation_boundary.ld_key = m_logdev->get_last_flush_ld_key(); }
+    if (req->seq_num == 0) {
```

Comment:
I dont find where we update `m_safe_truncation_boundary.seq_num` ? How this is expected to be update?

Reply:
yes I didnt follow this logic of this part ,  what is `m_safe_truncation_boundary.ld_key` and why it need to be updated in every write while the seq_num only update in truncate?

```
    /**
     * @brief Get the last truncated seqnum upto which we have truncated. If called after recovery, it returns the
     * first seq_num it has seen-1.
     *
     * @return logstore_seq_num_t
     */
    logstore_seq_num_t truncated_upto() const {
        const auto ts{m_safe_truncation_boundary.seq_num.load(std::memory_order_acquire)};
        return (ts == std::numeric_limits< logstore_seq_num_t >::max()) ? -1 : ts;
    }
```
It seems like we use `m_safe_truncation_boundary.seq_num` to track last truncate position ?

Reply:
Yes. The `truncate` happens at both logstore and log_dev layer.
1. The real truncate happens at log_dev (the frees disk space) and it only understand `ld_key` saved at  `m_safe_truncation_boundary.ld_key` (refer to `LogDev::truncate(...)`), at the post truncate processing, it will be updated for every logstore of their safe boundary. 
2. in log store layer, it has an internal data structure for all the log records `m_records` and it only understands `seq_num` and when truncation happens, it also needs to truncate this data structure to free the log records (refer to: `HomeLogStore::do_truncate(...)`).